torch/distributed/_shard/sharded_tensor/_ops/matrix_ops.py

```diff
@@ -1,76 +1,328 @@
+    if narrow_idx is not None:
+        rank_idx = None
+        for idx, placement in enumerate(st._sharding_spec.placements):  # type: ignore[attr-defined]
+            if placement.rank() == current_rank:  # type: ignore[index]
+                rank_idx = idx  # type: ignore[attr-defined]
+        shard_metadata = st.metadata().shards_metadata[rank_idx]  # type: ignore[index]
+        mask = mask.narrow(
+            narrow_idx,
+            shard_metadata.shard_offsets[sharding_dim],
+            shard_metadata.shard_sizes[sharding_dim],
+        )
+    local_tensor = st.local_tensor().masked_fill(mask, value)
+    return local_tensor, st.sharding_spec(), st.size()
+
 
-    # Validate types
-    if not isinstance(input, ShardedTensor):
-        raise TypeError("input needs to be a ShardedTensor")
-    local_tensor = input.local_tensor()
-    local_tensor = torch.nn.functional.dropout(local_tensor, p=p)
-    return ShardedTensor._init_from_local_tensor(
-        local_tensor.contiguous(),
-        input.sharding_spec(),
-        input.size(),  # type: ignore[arg-type]
-        process_group=pg,
+_register_sharded_op_on_local_tensor(
+    torch.Tensor.masked_fill,
+    extra_check=sharded_masked_fill_check,
+    customized_func=sharded_masked_fill,
+)
+
+def sharded_view_check(*args, **kwargs):
+    """
+    Perform extra checks for the sharded_view op.
+
+    Args: same as ``torch.Tensor.view``.
+
+    Return: None
+    """
+    st = args[0]
+    shapes = args[1:]
+    if len(shapes) == 0:
+        raise ValueError("Missing *shape for sharded view op.")
+    if len(shapes) <= st.sharding_spec().dim:
+        raise NotImplementedError(
+            f"Shape having dim {len(shapes)} is not supported "
+            f"for sharded tensor sharded on dim {st.sharding_spec().dim}."
+        )
+    st_size = math.prod(st.size())
+    shape_size = math.prod(shapes)
+    if (
+        shape_size > st_size
+        or st_size % shape_size
+        or shapes.count(lambda x: x < -1) > 0
+    ):
+        raise ValueError("Shape is invalid for sharded tensor size.")
+    if shapes.count(-1) > 1:
+        raise ValueError("Only one dimension can be inferred for sharded view op.")
+
+
+def sharded_view(args, kwargs, pg):
+    """
+    Handles ``__torch_function__`` dispatch for the sharded_view op.
+    For now we always keep the sharding dim after view. For example, if
+    a sharded tensor with size [16, 5] and sharded by 0. If we now view
+    it as [4, 2, 2, 5], it will still be sharded by dim 0.
+
+    Args: same as ``torch.Tensor.view``.
+
+    Return:
+        local_tensor (Tensor): New local tensor to build the sharded tensor.
+        sharding_spec (:class:`torch.distributed._shard.sharding_spec.ShardingSpec`):
+            sharding spec of the new sharded tensor.
+        new_st_size (torch.Size): Size of the new sharded tensor.
+    """
+    st = args[0]
+    shapes = args[1:]
+    # Infer the dim which is specified with -1.
+    if shapes.count(-1):
```

Comment:
nit: Maybe just use `shapes.index(-1)` here and catch and ignore the exception instead of first using count then index.

Reply:
Ah.. I see. Let me change it.